FEATURE: Add ``@apply`` meta-keyword to the Fusion-language

mficzel opened the pull request:
``@apply`` allows to override multiple properties of a fusion-prototype with a single expression. This is useful when complex data structures are mapped to fusion prototypes.

The example shows the rendering of a ``teaserList``-array by using a Teaser-Component and passing all keys from each teaser to the fusion Object.

```
teasers = Neos.Fusion:Collection {
	collection = ${teaserList}
	itemName = 'teaser'
	itemRenderer = Vendor.Site:Teaser {
		@apply.teasers = ${teaser}
	}
}
```

The code avoids passing down each fusion-property explicitly to the child component. A similar concept with different syntax from the JavaScript world is known as ES6-Spreads.

Another use-case is to use ``Neos.Fusion:Renderer`` to render a prototype while type and properties are based on data from the context.

```
example = Neos.Fusion:Renderer {
	type = ${data.type}
	element.@apply.properties = ${data.properties}
}
```

That way some meta-programming can used in fusion and both prototype and properties are decided late in the rendering by the fusion runtime.

The keys below ``@apply`` are evaluated before the fusion-object and the ``@context`` and are initialized. Each key below ``@apply`` must return a key-value map (values other than an array it are ignored). During the evaluation of each fusion-path the values from ``@apply`` are always checked first.

If a property is defined via ``@apply`` this value is returned without evaluating the fusionPath.

The ``@process`` and ``@if``-rules of the original fusion-key are still applied even if a value from ``@apply`` is returned.

Since ``@apply`` is evaluated first the overwritten values are already present during the evaluation of ``@context`` and will overlay the properties of ``this`` if they are accessed.

``@apply`` supports the same extended syntax and ordering as fusion processors and supports multiple keys. The evaluation order is defined via ``@position``, the keys that are evaluated last will override previously defined keys. This is also similar to the rules for ``@process``.

```
test = Vendor.Site:Prototype {
	@apply.contextValue {
		@position = 'start'
		expression = ${ arrayValueFromContext }
	}
	@apply.fusionObject {
		@position = 'end'
		expression = Neos.Fusion:RawArray {
			value = "altered value"
		}
	}
}
```

Other than ``@context`` ``@apply`` is only valid for a single fusion path, so when subpathes or children are rendered they are not affected by the parents ``@apply`` unless they are explicitly passed down.

skurfuerst commented:
awesome <3

jonnitto commented:
After a  few messages with @mficzel I understand the concept and usage of `@spread`.  
**Very nice, thank you for pushing things forward!**

One critical aspect here is the documentation. Features like this need to be very well explained, perhaps with a small repo with some demo code.

mficzel commented:
@jonnitto the documentation is the last point on the checklist. I want to have the behavior sorted out before i write stuff. Currently i am nut sure regarding the following points.

1. What is the expected behavior of @if together with spreads
2. How to proceed with ArrayFusion Objects. Right now they only know about the keys they got from fusion and use spreads to override those but will not notice new keys introduced via spread.

kitsunet commented on Neos.Fusion/Classes/Core/Runtime.php:
```diff
@@ -399,25 +440,25 @@ protected function evaluateInternal($fusionPath, $behaviorIfPathNotFound, $conte
             if ($this->hasExpressionOrValue($fusionConfiguration)) {
                 return $this->evaluteExpressionOrValueInternal($fusionPath, $fusionConfiguration, $cacheContext, $contextObject);
```
Should spreads also somehow be involved here? I seem to miss how a spread would affect these values if not given to the fusion object but evaluated "lazily". "setPropertiesOnFusionObject" is only called for array type fusion objects. Anything else will just internally evaluate fusion paths via runtime and probably miss the spreads? Might need to see if I can construct a test case. I am also not sure if there isn't an easier way than a separate spread stack, but that might actually be the only way...

mficzel replied:
The spreads are checked earlier in evaluateInternal and thus this code is not reached if a spread value for this key was present.

mficzel replied:
The setFusionProperties was actually implemented at last and only notifies the array fusion objects about available keys. Even the array objects fetch the actual values via evaluate.

mficzel commented:
@beheist, the documentation is adjusted and I tried to explain a bit more. Can you check wether this is better now?

 However the fusion side is not a feature I see many integrators use so this will be a bit technical and mostly an edge case in the fusion code add. The es6 spreads in afx that will be built on top of this are another story, so this will be the key feature for integrators while the pr here is “just” the foundation.

jonnitto commented:
I think it would be nice to have also an example in the documentation where you compare it to the classical way. A kind of: Instead of writing this (example) you can shorten things up like this (example)

mficzel commented:
@jonnitto @beheist regarding documentation i suggest to keep the technical documentation precise and maybe a bit dry. The more detailed explanation about the advantages should be done in the release post. Even in this case i still suggest to keep the description of @properties short but emphasize the afx-spreads as integrator feature much more prominently.

I would like to avoid this being recognized as THE new way of assigning fusion properties wich it is not it just solves very specific cases many users will never run into.

I definitely plan to have the afx-spreads ready for the 4.2 release and will provide sort user-friendly explanations for the release announcement for all my pr's.

mficzel commented:
@skurfuerst @hlubek @grebaldi i adjusted the code and documentation to `@apply` and also fixed the nitpicks of Sebastian. I think this is ready to review again.

mficzel commented:
Thank you, both. 

I now removed the checklist from the description since it can easyly end up in the changelog otherwise.

beheist commented:
Hey @mficzel, sorry I didn't see the ping in time. I like the docs much better now, it's clear what the feature is for and how to use it. I would add some more context and a before/after example, but that can be added later too.
